include/phasar/DataFlow/PathSensitivity/PathSensitivityManagerMixin.h

```diff
@@ -0,0 +1,365 @@
+
+#include "phasar/DataFlow/IfdsIde/SolverResults.h"
+#include "phasar/DataFlow/PathSensitivity/ExplodedSuperGraph.h"
+#include "phasar/DataFlow/PathSensitivity/PathSensitivityConfig.h"
+#include "phasar/DataFlow/PathSensitivity/PathSensitivityManagerBase.h"
+#include "phasar/DataFlow/PathSensitivity/PathTracingFilter.h"
+#include "phasar/PhasarLLVM/Utils/LLVMIRToSrc.h"
+#include "phasar/PhasarLLVM/Utils/LLVMShorthands.h"
+#include "phasar/Utils/DFAMinimizer.h"
+#include "phasar/Utils/GraphTraits.h"
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SetVector.h"
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include <filesystem>
+#include <system_error>
+#include <type_traits>
+
+namespace psr {
+template <typename Derived, typename AnalysisDomainTy, typename GraphType>
+class PathSensitivityManagerMixin {
+  using n_t = typename AnalysisDomainTy::n_t;
+  using d_t = typename AnalysisDomainTy::d_t;
+
+  using NodeRef = typename ExplodedSuperGraph<AnalysisDomainTy>::NodeRef;
+  using graph_type = GraphType;
+  using graph_traits_t = GraphTraits<graph_type>;
+  using vertex_t = typename graph_traits_t::vertex_t;
+
+  struct PathsToContext {
+    llvm::DenseMap<size_t, unsigned> Cache;
+    llvm::SetVector<unsigned, llvm::SmallVector<unsigned, 0>> CurrPath;
+  };
+
+  static const ExplodedSuperGraph<AnalysisDomainTy> &
+  assertNotNull(const ExplodedSuperGraph<AnalysisDomainTy> *ESG) noexcept {
+    assert(ESG != nullptr && "The exploded supergraph passed to the "
+                             "pathSensitivityManager must not be nullptr!");
+    return *ESG;
+  }
+
+protected:
+  PathSensitivityManagerMixin(
+      const ExplodedSuperGraph<AnalysisDomainTy> *ESG) noexcept
+      : ESG(assertNotNull(ESG)) {
+    static_assert(std::is_base_of_v<PathSensitivityManagerBase<n_t>, Derived>,
+                  "Invalid CTRP instantiation: The Derived type must inherit "
+                  "from PathSensitivityManagerBase!");
+  }
+
+public:
+  template <
+      typename FactsRangeTy, typename ConfigTy,
+      typename Filter = DefaultPathTracingFilter,
+      typename = std::enable_if_t<is_pathtracingfilter_for_v<Filter, NodeRef>>>
+  [[nodiscard]] GraphType
+  pathsDagToAll(n_t Inst, FactsRangeTy FactsRange,
+                const PathSensitivityConfigBase<ConfigTy> &Config,
+                const Filter &PFilter = {}) const {
+    graph_type Dag;
+    PathsToContext Ctx;
+
+    for (const d_t &Fact : FactsRange) {
+      auto Nod = ESG.getNodeOrNull(Inst, std::move(Fact));
+
+      if (!Nod) {
+
+        // llvm::errs() << "At Inst " << llvmIRToString(Inst) << "; Fact: " <<
+        // Fact
+        //              << '\n';
+        // llvm::errs();
+
+        {
+          std::error_code EC;
+          llvm::raw_fd_ostream ROS("explicitesg-err.dot", EC);
+          ESG.printAsDot(ROS);
```

Comment:
Here we're dumping some stuff to a file with hardcoded name to the cwd. Shouldn't that rather be configurable or just printed to stderr?

Reply:
The problem here is that the explicit ESG frequently grows quite large and the commandline buffer might not be sufficient to show all output. So we write the ESG into a file instead. This happens in a fatal-error path only, so I don't think, it is a huge issue. 

Maybe we should write to a temp file instead? What do you think?

Reply:
Yes, that might be better. Then we should also print the file name.

Reply:
done